# 46장 제너레이터와 async/await


## 목차

- [46.1 제네레이터란?](#46.1)
- [46.2 제네레이터 함수의 정의](#46.2)
- [46.3 제네레이터 객체](#46.3)
- [46.4 제네레이터의 일시 중지와 재개](#46.4)
- [46.5 제네레이터의 활용](#46.5)
- [46.6 async/await](#46.6)


## 46.1 제네레이터란?<a name="46.1"></a>

- 제너레이터 : 코드 블록의 실행을 일시 중지했다가 필요한 시점에 재개할 수 있는 특 수한 함수

- 제너레이터 vs. 일반함수
	1. 제너레이터 함수는 함수 호출자에게 함수 실행의 제어권 양도 가능
		- 함수 제어권을 함수가 독점하는 것이 아닌 함수 호출자에게 양도 가능
	2. 제너레이터 함수는 함수 호출자와 함수의 상태를 주고받을 수 있음
		- 제너레이터 함수는 함수 호출자에게 상태를 전달할 수 있고 함수 호출자로부터 상태를 전달받을 수도 있다
	3. 제너레이터 함수를 호출하면 제너레이터 객체 반환함
		- 제너레이터 함수를 호출하면 함수 코드를 실행하는 것이 아니라 이터러블이면서 동시에 이터레이터인 제너레이터 객체를 반환함


## 46.2 제네레이터 함수의 정의<a name="46.2"></a>

- function* 키워드로 선언

- 하나 이상의 yield 표현식 포함

- 애스터리스크(*) 위치는 function 키워드와 함수 이름 사이면 어디든지 상관없음
	- 일관성 유지를 위해 function 키워드 바로 뒤에 붙이는 것을 권장

- 화살표 함수로 정의 불가능

- new 연산자와 함께 생성자 함수로 호출 불가능

```js
// 제너레이터 함수 선언문
function* genDecFunc() { 
	yield 1;
}

// 제너레이터 함수 표현식
const genExpFunc = function* () { 
	yield 1;
};

// 제너레이터 메서드
const obj = {
	* genObjMethod() {
		yield 1; 
	}
};

// 제너레이터 클래스 메서드
class MyClass {
	* genClsMethod() {
		yield 1; 
	}
}
```

```js
function* genFunc() { yield 1; } 
function * genFunc() { yield 1; } 
function *genFunc() { yield 1; } 
function*genFunc() { yield 1; }
```

## 46.3 제네레이터 객체<a name="46.3"></a>

- 제너레이터 함수를 호출하면 일반 함수처럼 함수 코드 블록을 실행하는 것이 아니라 제너레이터 객체를 생성해 반환

- 제너레이터 함수가 반환한 제너레이터 객체는 이터러블이면서 동시에 이터레이터이다.

- 제너레이터 객체는 next 메서드를 갖는 이터레이터이지만, 이터레이터에 없는 return, throw 메서드를 가짐
	- 제너레이터 객체의 세 개의 메서드를 호출하면, 제너레이터 함수의 yield 표현식까지 코드 블록을 실행하고 yield된 값을 value 프로퍼티 값으로 반환하고 false를 done 프로퍼티 값으로 갖는 이터레이터 리절트 객체 반환함
	- return 메서드를 호출하면 인수로 전달받은 값을 value 프로퍼티 값으로 true를 done 프로퍼티 값으로 갖는 이터레이터 리절트 객체를 반환함
	- throw 메서드를 호출하면 인수로 전달받은 에러를 발생시키고 undefined를 value 프로퍼티 값으로 true를 done 프로퍼티 값으로 갖는 이터레이터 리절트 객체를 반환함

```js
// 제너레이터 함수
function* genFunc() { 
	yield 1;
	yield 2;
	yield 3; 
}

// 제너레이터 함수를 호출하면 제너레이터 객체 반환
const generator = genFunc();

// 제너레이터 객체는 이터러블이면서 동시에 이터레이터
// 이터러블은 Symbol.iterator 메서드를 직접 구현 or 프로토타입 체인인을 통해 상속받은 객체
console.log(Symbol.iterator in generator); // true
// 이터레이터는 next 메서드 가짐
console.log(1 next' in generator); // true
```


## 46.4 제네레이터의 일시 중지와 재개<a name="46.4"></a>

- yield 키워드와 next 메서드를 통해 실행을 일시 중지했다가 필요한 시점에 다시 재개 가능
	- 일반 함수는 호출 이후 제어권을 함수가 독점하나, 제너레이터는 함수 호출자에게 제어권을 양도하여 필요한 시점에 함수 실행 재개 가능

- 코드 블록의 모든 코드를 일괄 실행하지 않고, yield 표현식까지만 실행함

- yield 키워드는 제너레이터 함수의 실행을 일시 중지시키거나 yield 키워드 뒤에 오는 표현식의 평가결과를 제너레이터 함수 호출자에게 반환함

- 제너레이터 객체의 next 메서드에는 인수를 전달 가능

- 제너레이터 객체의 next 메서드에 전달한 인수는 제너레이터 함수의 yield 표현식을 할당받는 변수에 할당됨

```js
// 제너레이터 함수
function* genFunc() { 
	yield 1;
	yield 2;
	yield 3; 
}

// 제너레이터 함수를 호출하면 제너레이터 객체 반환
// 이터러블이면서 동시에 이터레이터인 제너레이터 객체는 next 메서드 가짐 
const generator = genFunc();

// 처음 next 메서드를 호출하면 첫 번째 yield 표현식까지 실행되고 일시 중지
// next 메서드는 이터레이터 리절트 객체({value, done}) 반환
// value 프로퍼티에는 첫 번째 yield 표현식에서 yield된 값 1 할당
// done 프로퍼티에는 제너레이터 함수가 끝까지 실행되었는지를 나타내는 false 할당
console.log(generator.next()); // {value: 1, done: false}

// 다시 next 메서드를 호출하면 두 번째 yield 표현식까지 실행되고 일시 중지
// next 메서드는 이터레이터 리절트 객체({value, done}) 반환
// value 프로퍼티에는 두 번째 yield 표현식에서 yield된 값 2 할당
// done 프로퍼티에는 제너레이터 함수가 끝까지 실행되었는지를 나타내는 false가 할당
console.log(generator.next()); // {value: 2, done: false}

// 다시 next 메서드를 호출하면 세 번째 yield 표현식까지 실행되고 일시 중지
// next 메서드는 이터레이터 리절트 객체({value, done}) 반환
// value 프로퍼티에는 세 번째 yield 표현식에서 yield된 값 3 할당
// done 프로퍼티에는 제너레이터 함수가 끝까지 실행되었는지를 나타내는 false가 할당
console.log(generator.next()); // {value: 3, done: false}

// 다시 next 메서드를 호출하면 남은 yield 표현식이 없으므로 제너레이터 함수의 마지막까지 실행
// next 메서드는 이터레이터 리절트 객체({value, done}) 반환
// value 프로퍼티에는 제너레이터 함수의 반환값 undefined 할당
// done 프로퍼티에는 제너레이터 함수가 끝까지 실행되었음을 나타내는 true 할당
console.log(generator.next()); // {value: undefined, done: true}
```

```js
function* genFunc() {
	// 처음 next 메서드를 호출하면 첫 번째 yield 표현식까지 실행되고 일시 중지
	// 이때 yield된 값 1은 next 메서드가 반환한 이터레이터 리절트 객체의 value 프로퍼티에 할당
	// x 변수에는 아직 할당되지 않음 
	// x 변수의 값은 next 메서드가 두 번째 호출될 때 결정
	const x = yield 1;

	// 두 번째 next 메서드를 호출할 때 전달한 인수 10은 첫 번째 yield 표현식을 할당받는 x 변수에 할당됨
	// const x = yield 1;은 두 번째 next 메서드를 호출했을 때 완료됨
	// 두 번째 next 메서드를 호출하면 두 번째 yield 표현식까지 실행되고 일시 중지됨
	// yield된 값 x + 10은 next 메서드가 반환한 이터레이터 리절트 객체의 value 프로퍼티에 할당됨
	const y = yield (x + 10);

	// 세 번째 next 메서드를 호출할 때 전달한 인수 20은 두 번째 yield 표현식을 할당받는 y 변수에 할당됨
	// const y = yield (x + 10);는 세 번째 next 메서드를 호출했을 때 완료됨
	// 세 번째 next 메서드를 호출하면 함수 끝까지 실행됨
	// 이때 제너레이터 함수의 반환값 x + y는 next 메서드가 반환한 이터레이터 리절트 객체의 value 프로퍼티에 할당됨
	// 일반적으로 제너레이터의 반환값은 의미 없음
	// 따라서 제너레이터에서는 값을 반환할 필요가 없고 return은 종료의 의미로만 사용해야 함 
	return x + y;
}

// 제너레이터 함수를 호출하면 제너레이터 객체 반환
// 이터러블이며 동시에 이터레이터인제너레이터객체는 next 메서드 가짐
const generator = genFunc(0);

// 처음 호출하는 next 메서드에는 인수를 전달하지 않음
// 만약 차음 호출하는 next 메서드에 인수를 전달하면 무시
// next 메서드가 반환한 이터레이터 리절트 객체의 value 프로퍼티에는 첫 번째 yield된 값 1이 할당됨 
let res = generator.next();
console.log(res); // {value: 1, done: false}

// next 메서드에 인수로 전달한 10은 genFunc 함수의 x 변수에 할당
// next 메서드가 반환한 이터레이터 리절트 객체의 value 프로퍼티에는 두 번째 yield된 값 20이 할당됨 
res = generator.next(10);
console.log(res); // {value: 20, done: false}

// next 메서드에 인수로 전달한 20은 genFunc 함수의 y 변수에 할당
// next 메서드가 반환한 이터레이터 리절트 객체의 value 프로퍼티에는 제너레이터 함수의 반환값 30이 할당
res = generator.next(20);
console.log(res); // {value: 30, done: true}
```


## 46.5 제네레이터의 활용<a name="46.5"></a>

### 46.5.1 이터러블의 구현

```js
// 무한 이터러블을 생성하는 제너레이터 함수
const infiniteFibonacci = (function* () { 
	let [pre, cur] = [0, 1];
	
	while (true) {
		[pre, cur] = [cur, pre + cur]; 
		yield cur;
	} 
}());

// infiniteFibonacci는 무한 이터러블
for (const num of infiniteFibonacci) {
	if (num > 10000) break;
	console.log(num); // 1 2 3 5 8... 2584 4181 6765 
}
```

### 46.5.2 비동기 처리

```js
const fetch = require('node-fetch'); 
// https://github.com/tj/co
const co = require('co');

co(function* fetchTodo() {
	const url = 'https://jsonplaceholder.typicode.com/todos/1';
	
	const response = yield fetch(url);
	const todo = yield response.json();
	console.log(todo);
	// { userid: 1, id: 1, title: 'delectus aut autem', completed: false }
});
```


## 46.6 async/await<a name="46.5"></a>

- ES8에서 제너레이터보다 간단하고 가독성 좋게 비동기 처리를 동 기 처리처럼 동작하도록 구현할 수 있는 async/await가 도입됨

- async/await는 프로미스를 기반으로 동작

- 프로미스의 후속 처리 메서드 없이 마치 동기 처리처럼 프로미스가 처리 결과를 반환하도록 구현 가능

```js
const fetch = require('node-fetch');

async function fetchTodo() {
	const url = 'https://jsonplaceholder.typicode.com/todos/1';

	const response = await fetch(url); 
	const todo = await response.json(); 
	console.log(todo);
	// {userid: 1, id: 1, title: 'delectus aut autem', completed: false}
}

fetchTodo();
```

### 46.6.1 async 함수

- await 키워드는 반드시 async 함수 내부에서 사용해야 함

- async 함수는 async 키워드를 사용해 정의하며 언제나 프로미스를 반환

- async 함수가 명시적으로 프로미스를 반환하지 않더라도 async 함수는 암묵적으로 반환값을 resolve하는 프로미스를 반환함

```js
// async 함수 선언문
async function foo(n) { return n; } 
foo(l).then(v => console.log(v)); // 1

// async 함수 표현식
const bar = async function (n) { return n; }; 
bar(2).then(v => console.log(v)); // 2

// async 화살표 함수
const baz = async n => n; 
baz(3).then(v => console.log(v)); // 3

// async 메사드 
const obj = {
	async foo(n) { return n; } 
};
obj.foo(4).then(v => console.log(v)); // 4

// async 클래스 메서드 
class MyClass {
	async bar(n) { return n; } 
}
const myClass = new MyClass(); 
myClass.bar(5).then(v => console.log(v)); // 5
```

```js
// 클래스의 contructor 메서드는 asyn 메서드가 될 수 없음
class MyClass {
	async constructor() { }
	// SyntaxError: Class constructor may not be an async method
}

const myClass = new MyClass();
```

### 46.6.2 await 키워드

- await 키워드는 프로미스가 settled 상태(비동기 처리가 수행된 상태)가 될 때까지 대기하다가 settled 상태가 되면 프로미스가 resolve한 처리 결과 반환 

- await 키워드는 반드시 프로미스 앞에서 사용해야 함

```js
const fetch = require('node-fetch');
const getGithubUserName = async id => {
	const res = await fetch('https://api.github.com/users/${id} ); // 1 
	const { name } = await res.json(); // 2
	console.log(name); // Ungmo Lee
};

getGithubUser_e('ungmo2');
```

```js
async function bar(n) {
	const a = await new Promise(resolve => setTimeout(() => resolve(n), 3000));
	// 두 번째 비동기 처리를 수행하려면 첫 번째 비동기 처리 결과가 필요
	const b = await new Promise(resolve => setTimeout(() => resolve(a + 1), 2000)); // 세 번째 비동기 처리를 수행하려면 두 번째 비동기 처리 결과가 필요
	const c = await new Promise(resolve => setTimeout(() => resolve(b + 1), 1000));
	
	console.log([a, b, c]); // [1, 2, 3] 
}

bar(l); // 약 6초
```

### 46.6.3 에러 처리

- try... catch 문을 사용 가능

- 프로미스를 반환하는 비동기 함수는 명시적으로 호출할 수 있기 때문에 호출자가 명확함

- async 함수 내에서 catch 문을 사용해서 에러 처리를 하지 않으면 async 함수는 발생한 에러를 reject하는 프로미스 반환

```js
const fetch = require('node-fetch');

const foo = async () => {
	try {
		const wrongUrl = 'https://wrong.url';
		const response = await fetch(wrongUrl); 
		const data = await response.json(); 
		console.log(data);
	} catch (err) {
		console.error(err); // TypeError: Failed to fetch
	} 
};

foo();
```

```js
const fetch = require('node-fetch');
const foo = async () => {
	const wrongUrl = 'https://wrong.url';
	
	const response = await fetch(wrongUrl); 
	const data = await response.json(); 
	return data;
};

foo()
	.then(console.log)
	.catch(console.error); // TypeError: Failed to fetch
```